RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,336 @@
+* If `e` is `empty`, it evaluates to `empty`.
+* If `e` is `undefined`, it evaluates to `undefined`.
+
+This definition keeps `der(e)` in symbolic form in a model that does not contain any units.
+Otherwise, `der(x) = x` would lead to unit errors even if `x` has no declared unit.
+This also means that in `k*der(x) = -x`, the unit of `k` won't be inferred to be `"s"`.
+
+### Evaluation
+Unit variables can't be found to be `empty` or `undefined` unit.
+This means that after evaluation, a unit meta-expression is either `undefined`, `empty` or does not contain any `undefined` or `empty` unit.
+
+After evaluation a constraint will not contain both a well-formed unit and a `der` meta-expression.
+This means that all constraints can be simplified to have at most one well-formed unit.
+
+## Unit equivalence and convertibility
+
+### Equivalence
+Two well-formed units are said to be equivalent if both of the following are true:
+* Their base unit factorizations are equal.
+* Their base unit scalings are equal.
+
+Examples:
+- `"N"` and `"m/s2"` have different base unit factorizations, they are not equivalent.
+- `"s"` and `"ms"` have different base unit scaling, they are not equivalent.
+- `"K"` and `"degC"` are equivalent.
+- `"kN"` and `"W.s/mm"` are equivalent.
+
+It is a quality of implementation how equality of scalings are checked.
+
+`empty` units and `undefined` units are always equivalent to each other, themselves and any `well-formed` units.
+
+### Convertibility
+A well-formed unit `u1` is said to be convertible to another well-formed unit `u2` if:
+* Their base unit factorizations are equal.
+
+Examples:
+- `"N"` is not convertible to `"m/s2"`, since they have different base unit factorizations.
+- `"s"` is convertible to `"ms"`, since they differ by base unit scaling.
+- `"K"` is convertible to `"degC"`.
+- `"kN"` is convertible to `"W.s/mm"`.
+
+`empty` units and `undefined` units are never convertible to each other, themselves or any `well-formed` units.
+
+## Unit inference
+
+### Modified Hindley-Milner
+
+The original Hindley-Milner algorithm is designed to infer types, and consists of building a set of substitutions that returns the type of any expression of a given program.
+Applying this algorithm to perform unit inference in Modelica comes with two important differences:
+  * Not all variables must have a unit in a given model.
+  * There are operations on meta-expressions, that is, not all constraints are of the form  _`var1.unit` is equivalent to `var2.unit`_.
+
+#### Solvability of constraints
+
+After meta-expression evaluation and gathering of variables in the single-sided form, a constraint's solvability can be determined using:
+* A set `V` of unit variables present with non-zero exponent.
+* A set `D` of unit variables, present inside a `der` meta-expression.
+
+With this information, a variable `var.unit` can be solved from a given constraint if:
+* `var.unit` is present in `V`, but not in `D`.
```

Comment:
I understand that this was designed to handle `der(x)=-x;` without reporting errors.

However, as far as I understand that also means that if we have `der(x)=velocity;` we cannot deduce the unit of `x`, which would be a step backwards.  To me it instead makes sense to state that only variables that occur exactly once can be solved from the constraint - which gives a similar result for this `der(x)=-x;`, but handles `der(x)=velocity;`.

The disadvantage is that we can then not deduce the unit of x in `x*x=area;` or `x+x=area;`. However, such examples are bit silly, especially as they can be written as `x^2=area;` or `2*x=area;`.

Reply:
> #### der
> Consider the meta-expression `der(e)`.
> * In a constraint where a `well-formed` unit is present, it evaluates to `e * ("s" ^ -1)`.
> [...]
> ### Evaluation
> [...]
> After evaluation a constraint will not contain both a well-formed unit and a `der` meta-expression.
...

Feel free to rephrase so that it is clearer.
At the end of the day the complexity associated with relaxing unit-checking is why we felt the need to spell out the algorithm in painstaking details in this document.

Reply:
My proposal is not rephrase it, but instead state that only variables that only occur once in a constraint can have their unit inferred.

Reply:
Which would make the following invalid right?
```
model Test3
  Real x;
  Real y;
  equation
  der(x) = y;
  y = -x;
end Test3;
```

Reply:
> Which would make the following invalid right?
> 
> ```
> model Test3
>   Real x;
>   Real y;
>   equation
>   der(x) = y;
>   y = -x;
> end Test3;
> ```

No, since we cannot start the process.
Similarly as not being able to deduce units for `x` and `y` based on:
 `sin(x/y)=...; cos(y*x)=...; `

Reply:
In the language of this proposal, you would want for the unit of `sin`, `cos` and `der` to be `undefined`?

Reply:
> In the language of this proposal, you would want for `sin` and `cos` to not impose any constraint on their argument, as well as expressions containing `der` to have `undefined` unit?

I would rather that:
-  `sin` and `cos` impose unit="1" (or "rad") for input.
-  `sin` and `cos` impose unit="1" for output.
- der(x) impose that unit(der(x))=unit(x) * "s-1" (or something like that)

Reply:
What would you conclude for the following mode:
```
model Test4
  Real x;
  Real y;
  Real c(unit = "m");
equation
  c = x + x*y;
end Test4;
```

Reply:
Yes, and x.unit="m" (based on c=x... where c.unit="m") and y.unit="1" (based on c=...+x*y; where c.unit="m" and x.unit="m"). I agree that it takes some care to specify the constraints that unit(a+b)=unit(a)=unit(b); to not get stuck on unit for x*y.